File: model-optimizer/mo/front/common/partial_infer/tile.py
"""
 Copyright (c) 2018 Intel Corporation

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""

import logging as log

import numpy as np


def tf_tile_infer(node):
    shape = node.in_node(0).shape
    if shape is None:
        log.error("Undefined shape for the input tiles for the Tile operation '{}'.".format(node.node))
        return

    tile_array = node.in_node(1).value
    assert (len(shape) == len(tile_array))
    non_one_tile = np.argwhere(tile_array != 1)
    if len(non_one_tile) != 1:
        node['type'] = None
        log.warning("Tile operation with more than one dimension not equal to 1 is not supported")
    elif len(non_one_tile) == 1:
        node['axis'] = non_one_tile[0][0]
        node['tiles'] = tile_array[node['axis']]
        node.graph.remove_edge(node.in_node(1).id, node.id)

    node.out_node().shape = shape * tile_array
    if node.in_node(0).value is not None:
        node.out_node().value = np.tile(node.in_node(0).value, tile_array)

File: ngraph/test/type_prop/tile.cpp
//*****************************************************************************
// Copyright 2017-2020 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//*****************************************************************************

#include "gtest/gtest.h"
#include "ngraph/ngraph.hpp"
#include "util/type_prop.hpp"

using namespace std;
using namespace ngraph;

TEST(type_prop, tile)
{
    auto param0 = make_shared<op::Parameter>(element::f32, Shape{6, 8, 10});
    auto param1 = op::Constant::create(element::i64, Shape{3}, {3, 4, 1});
    auto top = make_shared<op::v0::Tile>(param0, param1);
    ASSERT_EQ(top->get_element_type(), element::f32);
    ASSERT_EQ(top->get_shape(), (Shape{18, 32, 10}));
}

TEST(type_prop, tile_small_data_rank)
{
    auto param0 = make_shared<op::Parameter>(element::f32, Shape{8, 10});
    auto param1 = op::Constant::create(element::i64, Shape{3}, {3, 4, 1});
    auto top = make_shared<op::v0::Tile>(param0, param1);
    ASSERT_EQ(top->get_element_type(), element::f32);
    ASSERT_EQ(top->get_shape(), (Shape{3, 32, 10}));
}

TEST(type_prop, tile_few_repeats)
{
    auto param0 = make_shared<op::Parameter>(element::f32, Shape{6, 8, 10});
    auto param1 = op::Constant::create(element::i64, Shape{2}, {4, 1});
    auto top = make_shared<op::v0::Tile>(param0, param1);
    ASSERT_EQ(top->get_element_type(), element::f32);
    ASSERT_EQ(top->get_shape(), (Shape{6, 32, 10}));
}
